utils and libraries: Make 'blocksdir' always net specific

hebasto opened the pull request:
The blocks directory is net specific by definition.

Also this prevents the side effect of calling `GetBlocksDir(false)` in the non-mainnet environment.
Currently a new node creates an unused `blocks\` directory in the root of the data directory when `-testnet` or `-regtest` is specified.

Refs:
- #12653
- https://github.com/bitcoin/bitcoin/pull/12653#discussion_r174784834 by @laanwj 
- https://github.com/bitcoin/bitcoin/issues/14595#issuecomment-436011186

hebasto committed:
Make blockdir always net specific

The blocks directory is net specific by definition.

Also this prevents the side effect of calling GetBlocksDir(false) in the
non-mainnet environment.

ken2812221 commented:
Concept ACK. This seem to require a release note.

MarcoFalke commented:
Needs tests before merge

MeshCollider commented:
Concept ACK, so this only affects the case where the user has explicitly specified a directory with the `-blocksdir` argument right?

hebasto commented:
@MeshCollider 

Thank you for your review.

> Concept ACK, so this only affects the case where the user has explicitly specified a directory with the `-blocksdir` argument right?

This PR works regardless of the `-blocksdir` option.

sipa commented:
Does this risk breaking any existing configurations?

MeshCollider commented:
@sipa that was my concern, the actual behaviour looks unchanged but I think now if you specify a base directory with `-blocksdir` it will error if the child directory doesn't exist, previously it would just create it?

hebasto commented:
@MeshCollider @sipa 

> ... I think now if you specify a base directory with `-blocksdir` it will error if the child directory doesn't exist, previously it would just create it? 

Test of this PR:
```
$ rm -rf /home/hebasto/.bitcoin/
$ /home/hebasto/github/bitcoin/src/bitcoind
2018-10-13T09:27:14Z Bitcoin Core version v0.17.99.0-c1a60d4a1 (release build)
...
2018-10-13T09:27:30Z Shutdown: done
$ find /home/hebasto/.bitcoin/ > ~/pr14409-without-blocksdir

$ rm -rf /home/hebasto/.bitcoin/
$ /home/hebasto/github/bitcoin/src/bitcoind -blocksdir=/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
2018-10-13T09:33:59Z Bitcoin Core version v0.17.99.0-c1a60d4a1 (release build)
2018-10-13T09:33:59Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2018-10-13T09:33:59Z Error: Specified blocks directory "/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR" does not exist.
Error: Specified blocks directory "/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR" does not exist
$ mkdir -p /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
$ /home/hebasto/github/bitcoin/src/bitcoind -blocksdir=/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
2018-10-13T09:36:37Z Bitcoin Core version v0.17.99.0-c1a60d4a1 (release build)
...
2018-10-13T09:37:00Z Shutdown: done
$ find /home/hebasto/.bitcoin/ > ~/pr14409-with-blocksdir

$ diff ~/pr14409-without-blocksdir ~/pr14409-with-blocksdir 
3a4,8
> /home/hebasto/.bitcoin/TEST_DIR
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks/rev00000.dat
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks/blk00000.dat
12d16
< /home/hebasto/.bitcoin/blocks/rev00000.dat
18d21
< /home/hebasto/.bitcoin/blocks/blk00000.dat
```

Test of master be992701b:
```
$ rm -rf /home/hebasto/.bitcoin/
$ /home/hebasto/github/bitcoin/src/bitcoind
2018-10-13T10:19:33Z Bitcoin Core version v0.17.99.0-be992701b (release build)
...
2018-10-13T10:19:51Z Shutdown: done
$ find /home/hebasto/.bitcoin/ > ~/master-be992701b-without-blocksdir

$ rm -rf /home/hebasto/.bitcoin/
$ /home/hebasto/github/bitcoin/src/bitcoind -blocksdir=/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
2018-10-13T10:22:30Z Bitcoin Core version v0.17.99.0-be992701b (release build)
2018-10-13T10:22:30Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2018-10-13T10:22:30Z Error: Specified blocks directory "/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR" does not exist.
Error: Specified blocks directory "/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR" does not exist.
$ mkdir -p /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
$ /home/hebasto/github/bitcoin/src/bitcoind -blocksdir=/home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
2018-10-13T10:23:18Z Bitcoin Core version v0.17.99.0-be992701b (release build)
...
2018-10-13T10:23:35Z Shutdown: done
$ find /home/hebasto/.bitcoin/ > ~/master-be992701b-with-blocksdir

$ diff ~/master-be992701b-without-blocksdir ~/master-be992701b-with-blocksdir 
3a4,8
> /home/hebasto/.bitcoin/TEST_DIR
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks/rev00000.dat
> /home/hebasto/.bitcoin/TEST_DIR/TEST_BLOCKS_DIR/blocks/blk00000.dat
12d16
< /home/hebasto/.bitcoin/blocks/rev00000.dat
18d21
< /home/hebasto/.bitcoin/blocks/blk00000.dat
```

The behaviour is unchanged. If you specify a base directory with `-blocksdir` and if the child directory doesn't exist it will an error both on master and on this PR.

hebasto commented:
Rebased.

laanwj commented:
This absolutely needs testing in a functional test.

Also to make it clear what behavior is failing right now, and which will pass after this.

hebasto committed:
Improve blocksdir functional test.

A new node should not create an unused `blocks` directory in the root of
the data directory when `-testnet` or `-regtest` is specified.

laanwj commented:
utACK e4a0c3547ed886871f8b3d51c6b4ffdb181a8b9c

MarcoFalke commented:
utACK e4a0c3547ed886871f8b3d51c6b4ffdb181a8b9c

fanquake commented:
Being backported in #15445.
